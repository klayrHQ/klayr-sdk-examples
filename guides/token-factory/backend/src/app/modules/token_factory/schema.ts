export const createTokenSchema = {
	$id: 'token_factory/createToken-params',
	title: 'CreateTokenCommand transaction parameter for the Token Factory module',
	type: 'object',
	required: ['name'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export const configSchema = {
	$id: '/token_factory/config',
	type: 'object',
	properties: {
		maxNameLength: {
			type: 'integer',
			format: 'uint32',
		},
		maxSymbolLength: {
			type: 'integer',
			format: 'uint32',
		},
		maxTotalSupply: {
			type: 'integer',
			format: 'uint64',
		},
	},
	required: ['maxNameLength', 'maxSymbolLength', 'maxTotalSupply'],
};
