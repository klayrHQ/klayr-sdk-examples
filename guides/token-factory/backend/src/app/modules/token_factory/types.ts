import { JSONObject } from 'klayr-sdk';

export type TokenID = Buffer;

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: bigint;
	minAmountToMint: bigint;
	maxAmountToMint: bigint;
	chainID: Buffer;
	createTokenFee: bigint;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
