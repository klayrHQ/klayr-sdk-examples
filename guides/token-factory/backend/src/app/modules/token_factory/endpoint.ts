import { BaseEndpoint } from 'klayr-sdk';

export class TokenFactoryEndpoint extends BaseEndpoint {}
