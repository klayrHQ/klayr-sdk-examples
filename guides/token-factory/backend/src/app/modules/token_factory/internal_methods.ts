import { BaseMethod, CommandExecuteContext } from 'klayr-sdk';
import { OwnerStore } from './stores/owner';
import { CounterStore, CounterStoreData, counterKey } from './stores/counter';
import { TokenStore } from './stores/token';
import { CreateTokenParams } from './commands/create_token_command';

export class InternalMethod extends BaseMethod {
	private _chainID!: Buffer;

	public async init(chainID: Buffer): Promise<void> {
		this._chainID = chainID;
	}

	public async getNextTokenIDAndSetStores(
		context: CommandExecuteContext<CreateTokenParams>,
		senderAddress: Buffer,
	) {
		const tokenStore = this.stores.get(TokenStore);
		const counterStore = this.stores.get(CounterStore);
		const ownerStore = this.stores.get(OwnerStore);

		let tokenIdCounter: CounterStoreData = await counterStore
			.get(context, counterKey)
			.catch(() => ({ counter: 0 }));
		tokenIdCounter.counter++;

		const newLocalIDBuffer = Buffer.alloc(4);
		newLocalIDBuffer.writeUInt32BE(tokenIdCounter.counter);
		const newTokenIDBuffer = Buffer.concat([this._chainID, newLocalIDBuffer]);

		await Promise.all([
			counterStore.set(context, counterKey, tokenIdCounter),
			ownerStore.set(context, newTokenIDBuffer, { address: senderAddress }),
			tokenStore.set(context, newTokenIDBuffer, {
				name: context.params.name,
				symbol: context.params.symbol,
				totalSupply: BigInt(context.params.totalSupply),
			}),
		]);

		return newTokenIDBuffer;
	}
}
