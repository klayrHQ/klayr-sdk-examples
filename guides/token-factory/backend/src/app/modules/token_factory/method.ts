import { BaseMethod, CommandVerifyContext } from 'klayr-sdk';
import { OwnerStore, OwnerStoreData } from './stores/owner';

export class TokenFactoryMethod extends BaseMethod {
	public async getOwnerOfToken<T>(
		context: CommandVerifyContext<T>,
		tokenID: Buffer,
	): Promise<OwnerStoreData | undefined> {
		try {
			const ownerStore = this.stores.get(OwnerStore);
			return await ownerStore.get(context, tokenID);
		} catch {
			return undefined;
		}
	}
}
