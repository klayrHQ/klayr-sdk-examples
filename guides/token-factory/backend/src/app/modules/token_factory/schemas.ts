export const mintSchema = {
	$id: 'token_factory/mintParams',
	title: 'Mints extra tokens to a recipient',
	type: 'object',
	properties: {
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['amount', 'recipient'],
};

export const createTokenSchema = {
	$id: 'token_factory/createToken-params',
	title: 'CreateTokenCommand transaction parameter for the Token Factory module',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export const configSchema = {
	$id: '/token_factory/config',
	type: 'object',
	properties: {
		maxNameLength: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		maxSymbolLength: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		maxTotalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		minAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 4,
		},
		maxAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		createTokenFee: {
			dataType: 'uint64',
			fieldNumber: 6,
		},
	},
	required: [
		'maxNameLength',
		'maxSymbolLength',
		'maxTotalSupply',
		'minAmountToMint',
		'maxAmountToMint',
		'createTokenFee',
	],
};
