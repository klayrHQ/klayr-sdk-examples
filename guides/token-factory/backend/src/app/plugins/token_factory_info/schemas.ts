export const configSchema = {
	$id: '/plugins/token_factory/config',
	type: 'object',
	properties: {
		syncInterval: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		chainID: {
			dataType: 'string',
			fieldNumber: 2,
		},
	},
	required: ['syncInterval', 'chainID'],
	default: {
		syncInterval: 20000, // milliseconds
		chainID: '12345678',
	},
};

export const mintEventSchema = {
	$id: 'token_factory/mintEvent',
	type: 'object',
	properties: {
		address: {
			dataType: 'bytes',
			fieldNumber: 1,
			format: 'klayr32',
		},
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		result: {
			dataType: 'uint32',
			fieldNumber: 4,
		},
	},
	required: ['address', 'tokenID', 'amount', 'result'],
};

export const newTokenEventSchema = {
	$id: 'token_factory/new_token',
	title: 'New token created with the createToken command',
	type: 'object',
	required: ['owner', 'name', 'symbol', 'totalSupply', 'height'],
	properties: {
		owner: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		name: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 4,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 6,
		},
	},
};

export const heightSchema = {
	$id: '/token_factory/height',
	type: 'object',
	required: ['height'],
	properties: {
		height: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};
