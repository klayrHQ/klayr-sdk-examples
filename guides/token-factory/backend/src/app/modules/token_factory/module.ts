/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { validator } from '@klayr/validator';
import {
	BaseModule,
	FeeMethod,
	ModuleInitArgs,
	ModuleMetadata,
	TokenMethod,
	utils,
} from 'klayr-sdk';
import { BatchTransferCommand } from './commands/batch_transfer_command';
import { ModuleConfig, ModuleConfigJSON } from './types';
import { BurnCommand } from './commands/burn_command';
import { CreateTokenCommand } from './commands/create_token_command';
import { MintCommand } from './commands/mint_command';
import { TokenFactoryEndpoint } from './endpoint';
import { NewTokenEvent } from './events/new_token';
import { InternalMethod } from './internal_methods';
import { TokenFactoryMethod } from './method';
import { configSchema } from './schemas';
import { CounterStore } from './stores/counter';
import { OwnerStore } from './stores/owner';
import { TokenStore } from './stores/token';
import { getModuleConfig } from './utils';

export const defaultConfig = {
	maxNameLength: 30,
	maxSymbolLength: 5,
	maxTotalSupply: BigInt(1e18), // max: 9223372036854775807 = 9e18
	minAmountToMint: BigInt(1000),
	maxAmountToMint: BigInt(1e6) * BigInt(1e8),
	createTokenFee: BigInt(100_000),
	minAmountToBurn: BigInt(1000),
};

export class TokenFactoryModule extends BaseModule {
	private readonly _internalMethod = new InternalMethod(this.stores, this.events);

	private _createTokenCommand = new CreateTokenCommand(this.stores, this.events);
	private _batchTransferCommand = new BatchTransferCommand(this.stores, this.events);
	private _mintCommand = new MintCommand(this.stores, this.events);
	private _burnCommand = new BurnCommand(this.stores, this.events);

	private _moduleConfig!: ModuleConfig;
	private _tokenMethod!: TokenMethod;
	private _feeMethod!: FeeMethod;

	public endpoint = new TokenFactoryEndpoint(this.stores, this.offchainStores);
	public tokenFactoryMethod = new TokenFactoryMethod(this.stores, this.events);
	public commands = [
		this._createTokenCommand,
		this._mintCommand,
		this._burnCommand,
		this._batchTransferCommand,
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(TokenStore, new TokenStore(this.name, 0));
		this.stores.register(CounterStore, new CounterStore(this.name, 1));
		this.stores.register(OwnerStore, new OwnerStore(this.name, 2));

		this.events.register(NewTokenEvent, new NewTokenEvent(this.name));
	}

	public addDependencies(tokenMethod: TokenMethod, feeMethod: FeeMethod) {
		this._tokenMethod = tokenMethod;
		this._feeMethod = feeMethod;
		this._createTokenCommand.addDependencies({
			internalMethod: this._internalMethod,
			tokenMethod: this._tokenMethod,
			feeMethod: this._feeMethod,
		});
		this._mintCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._burnCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._batchTransferCommand.addDependencies({
			tokenMethod: this._tokenMethod,
		});
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			commands: this.commands.map(command => ({
				name: command.name,
				params: command.schema,
			})),
			events: this.events.values().map(v => ({
				name: v.name,
				data: v.schema,
			})),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(args: ModuleInitArgs): Promise<void> {
		// Get the module config defined in the config.json of the node
		const { moduleConfig, genesisConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the config with the config schema

		validator.validate<ModuleConfigJSON>(configSchema, config);

		this._moduleConfig = getModuleConfig(config, genesisConfig);
		this._internalMethod.init(this._moduleConfig.chainID);
		this._createTokenCommand.init(this._moduleConfig).catch(err => {
			console.log('Error: ', err);
		});
		this._mintCommand
			.init({
				minAmountToMint: this._moduleConfig.minAmountToMint,
				maxAmountToMint: this._moduleConfig.maxAmountToMint,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
		this._burnCommand
			.init({
				minAmountToBurn: this._moduleConfig.minAmountToBurn,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
	}

	// public async insertAssets(_context: InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {

	// }
}
