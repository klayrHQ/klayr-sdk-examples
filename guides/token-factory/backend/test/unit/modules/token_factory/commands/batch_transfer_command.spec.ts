/* eslint-disable */

import {
	BatchTransferCommand,
	Params,
} from '@app/modules/token_factory/commands/batch_transfer_command';
import { TokenFactoryModule } from '@app/modules/token_factory/module';
import { batchTransferParamsSchema as schema } from '@app/modules/token_factory/schemas';
import { TokenID, createCtx, createSampleTransaction } from '@test/helpers';
import {
	CommandExecuteContext,
	Transaction,
	VerifyStatus,
	chain,
	codec,
	cryptography,
	db,
} from 'klayr-sdk';

describe('BatchTransferCommand', () => {
	const tokenID = new TokenID(0).toBuffer();

	let batchTransfer: BatchTransferCommand;
	let stateStore: any;

	const getAvailableBalance = jest.fn();
	const mockTransfer = jest.fn();

	beforeEach(() => {
		const tokenFactory = new TokenFactoryModule();

		batchTransfer = new BatchTransferCommand(tokenFactory.stores, tokenFactory.events);
		batchTransfer.addDependencies({
			tokenMethod: { getAvailableBalance, transfer: mockTransfer } as any,
		});

		stateStore = new chain.StateStore(new db.InMemoryDatabase());

		// mock available balance in the sender wallet
		getAvailableBalance.mockResolvedValue(BigInt(1e12));
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(batchTransfer.name).toEqual('batchTransfer');
		});

		it('should have valid schema', () => {
			expect(batchTransfer.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			it('should throw errors for invalid schema', async () => {
				const paramWithInvalidAmount = codec.encode(schema, {
					tokenID: tokenID,
					amounts: [BigInt(1e14)],
					recipients: [cryptography.utils.getRandomBytes(20)],
				});

				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidAmount, batchTransfer.name),
				);
				const ctx = createCtx<Params>(stateStore, transaction, schema, 'verify');

				const result = await batchTransfer.verify(ctx);
				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error(`Insufficient Balance`));
			});

			it('should be ok for valid schema', async () => {
				const paramWithInvalidAmount = codec.encode(schema, {
					tokenID: tokenID,
					amounts: [BigInt(1e10)],
					recipients: [cryptography.utils.getRandomBytes(20)],
				});

				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidAmount, batchTransfer.name),
				);
				const ctx = createCtx<Params>(stateStore, transaction, schema, 'verify');

				const result = await batchTransfer.verify(ctx);
				expect(result.status).toBe(VerifyStatus.OK);
			});
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			it('should execute transfers', async () => {
				const amount = BigInt(5);
				const recipientAddress = cryptography.utils.getRandomBytes(20);

				const paramWithInvalidAmount = codec.encode(schema, {
					tokenID,
					amounts: [amount],
					recipients: [recipientAddress],
				});

				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidAmount, batchTransfer.name),
				);
				const ctx = createCtx<Params>(stateStore, transaction, schema, 'execute');
				await expect(
					batchTransfer.execute(ctx as CommandExecuteContext<Params>),
				).resolves.toBeUndefined();
				expect(mockTransfer).toHaveBeenCalledTimes(1);
			});
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
