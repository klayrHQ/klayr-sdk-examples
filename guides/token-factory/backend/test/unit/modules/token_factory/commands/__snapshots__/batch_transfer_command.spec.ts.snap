// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchTransferCommand constructor should have valid schema 1`] = `
{
  "$id": "token_factory/batchTransferParams",
  "properties": {
    "recipients": {
      "fieldNumber": 2,
      "items": {
        "properties": {
          "amount": {
            "dataType": "uint64",
            "fieldNumber": 2,
          },
          "recipient": {
            "dataType": "bytes",
            "fieldNumber": 1,
          },
        },
        "required": [
          "recipient",
          "amount",
        ],
        "type": "object",
      },
      "maxItems": 20,
      "minItems": 1,
      "type": "array",
    },
    "tokenID": {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
  },
  "required": [
    "tokenID",
    "recipients",
  ],
  "title": "Transfer transaction params",
  "type": "object",
}
`;
